File: src/bmsx_cpp/rompack/toc.cpp
#include "rompack/toc.h"

#include "common/endian.h"
#include "rompack/tokens.h"

#include <algorithm>
#include <unordered_map>
#include <utility>

namespace bmsx {
namespace {

struct TocStringSlice {
	u32 offset = ROM_TOC_INVALID_U32;
	u32 length = 0;
};

std::optional<i32> optionalI32FromU32(u32 value) {
	if (value == ROM_TOC_INVALID_U32) {
		return std::nullopt;
	}
	return static_cast<i32>(value);
}

std::optional<std::string> decodeTocString(const u8* table, size_t tableSize, u32 offset, u32 length) {
	if (offset == ROM_TOC_INVALID_U32 || length == 0) {
		return std::nullopt;
	}
	if (static_cast<size_t>(offset) + static_cast<size_t>(length) > tableSize) {
		throw BMSX_RUNTIME_ERROR("ROM TOC string table entry out of bounds.");
	}
	return std::string(reinterpret_cast<const char*>(table + offset), static_cast<size_t>(length));
}

u32 tocFieldValue(const std::optional<i32>& value) {
	return value.has_value() ? static_cast<u32>(*value) : ROM_TOC_INVALID_U32;
}

u32 tocUpdateLo(const std::optional<i64>& value) {
	return value.has_value() ? static_cast<u32>(static_cast<u64>(*value) & 0xffffffffu) : 0u;
}

u32 tocUpdateHi(const std::optional<i64>& value) {
	return value.has_value() ? static_cast<u32>(static_cast<u64>(*value) >> 32u) : 0u;
}

void writeTocString(std::vector<u8>& table, std::unordered_map<std::string, TocStringSlice>& index, std::string_view text, TocStringSlice& out) {
	if (text.empty()) {
		out = {};
		return;
	}
	const std::string key(text);
	const auto found = index.find(key);
	if (found != index.end()) {
		out = found->second;
		return;
	}
	out.offset = static_cast<u32>(table.size());
	out.length = static_cast<u32>(text.size());
	for (char value : text) {
		table.push_back(static_cast<u8>(value));
	}
	index.emplace(key, out);
}

TocStringSlice internTocString(std::vector<u8>& table, std::unordered_map<std::string, TocStringSlice>& index, const std::optional<std::string>& text) {
	TocStringSlice ref;
	if (text.has_value()) {
		writeTocString(table, index, *text, ref);
	}
	return ref;
}

TocStringSlice internTocString(std::vector<u8>& table, std::unordered_map<std::string, TocStringSlice>& index, const std::string& text) {
	TocStringSlice ref;
	writeTocString(table, index, text, ref);
	return ref;
}

} // namespace

std::string assetTypeFromId(u32 id) {
	switch (id) {
		case ROM_TOC_ASSET_TYPE_IMAGE: return "image";
		case ROM_TOC_ASSET_TYPE_AUDIO: return "audio";
		case ROM_TOC_ASSET_TYPE_DATA: return "data";
		case ROM_TOC_ASSET_TYPE_BIN: return "bin";
		case ROM_TOC_ASSET_TYPE_ATLAS: return "atlas";
		case ROM_TOC_ASSET_TYPE_ROMLABEL: return "romlabel";
		case ROM_TOC_ASSET_TYPE_MODEL: return "model";
		case ROM_TOC_ASSET_TYPE_AEM: return "aem";
		case ROM_TOC_ASSET_TYPE_LUA: return "lua";
		case ROM_TOC_ASSET_TYPE_CODE: return "code";
		default:
			throw BMSX_RUNTIME_ERROR("Unknown asset type id: " + std::to_string(id));
	}
}

u32 assetTypeToId(std::string_view type) {
	if (type == "image") return ROM_TOC_ASSET_TYPE_IMAGE;
	if (type == "audio") return ROM_TOC_ASSET_TYPE_AUDIO;
	if (type == "data") return ROM_TOC_ASSET_TYPE_DATA;
	if (type == "bin") return ROM_TOC_ASSET_TYPE_BIN;
	if (type == "atlas") return ROM_TOC_ASSET_TYPE_ATLAS;
	if (type == "romlabel") return ROM_TOC_ASSET_TYPE_ROMLABEL;
	if (type == "model") return ROM_TOC_ASSET_TYPE_MODEL;
	if (type == "aem") return ROM_TOC_ASSET_TYPE_AEM;
	if (type == "lua") return ROM_TOC_ASSET_TYPE_LUA;
	if (type == "code") return ROM_TOC_ASSET_TYPE_CODE;
	throw BMSX_RUNTIME_ERROR("Unknown asset type: " + std::string(type));
}

AssetTypeKind resolveAssetTypeKind(std::string_view assetType) {
	if (assetType.empty()) {
		return AssetTypeKind::Unknown;
	}
	switch (assetType[0]) {
		case 'i':
			if (assetType == "image") return AssetTypeKind::ImageAtlas;
			break;
		case 'a':
			if (assetType == "atlas") return AssetTypeKind::ImageAtlas;
			if (assetType == "audio") return AssetTypeKind::Audio;
			if (assetType == "aem") return AssetTypeKind::Aem;
			break;
		case 'm':
			if (assetType == "model") return AssetTypeKind::Model;
			break;
		case 'b':
			if (assetType == "bin") return AssetTypeKind::Bin;
			break;
		case 'l':
			if (assetType == "lua") return AssetTypeKind::Lua;
			break;
		case 'd':
			if (assetType == "data") return AssetTypeKind::Data;
			break;
		case 'r':
			if (assetType == "romlabel") return AssetTypeKind::Skip;
			break;
		case 'c':
			if (assetType == "code") return AssetTypeKind::Code;
			break;
	}
	return AssetTypeKind::Unknown;
}

RomTocPayload decodeRomToc(const u8* data, size_t size) {
	if (size < ROM_TOC_HEADER_SIZE) {
		throw BMSX_RUNTIME_ERROR("ROM TOC is too small.");
	}
	if (readLE32(data + 0) != ROM_TOC_MAGIC) {
		throw BMSX_RUNTIME_ERROR("Invalid ROM TOC magic.");
	}
	if (readLE32(data + 4) != ROM_TOC_HEADER_SIZE) {
		throw BMSX_RUNTIME_ERROR("Unexpected ROM TOC header size.");
	}
	const u32 entrySize = readLE32(data + 8);
	if (entrySize != ROM_TOC_ENTRY_SIZE) {
		throw BMSX_RUNTIME_ERROR("Unexpected ROM TOC entry size.");
	}
	const u32 entryCount = readLE32(data + 12);
	const u32 entryOffset = readLE32(data + 16);
	if (entryOffset != ROM_TOC_HEADER_SIZE) {
		throw BMSX_RUNTIME_ERROR("Unexpected ROM TOC entry offset.");
	}
	const u32 stringTableOffset = readLE32(data + 20);
	const u32 stringTableLength = readLE32(data + 24);
	const u32 projectRootOffset = readLE32(data + 28);
	const u32 projectRootLength = readLE32(data + 32);
	const size_t entriesBytes = static_cast<size_t>(entryCount) * static_cast<size_t>(entrySize);
	const size_t expectedStringOffset = static_cast<size_t>(entryOffset) + entriesBytes;
	if (static_cast<size_t>(stringTableOffset) != expectedStringOffset) {
		throw BMSX_RUNTIME_ERROR("Unexpected ROM TOC string table offset.");
	}
	if (static_cast<size_t>(entryOffset) + entriesBytes > size) {
		throw BMSX_RUNTIME_ERROR("ROM TOC entries out of bounds.");
	}
	if (static_cast<size_t>(stringTableOffset) + static_cast<size_t>(stringTableLength) > size) {
		throw BMSX_RUNTIME_ERROR("ROM TOC string table out of bounds.");
	}

	const u8* stringTable = data + stringTableOffset;
	const size_t stringTableSize = stringTableLength;
	RomTocPayload payload;
	payload.projectRootPath = decodeTocString(stringTable, stringTableSize, projectRootOffset, projectRootLength);
	payload.entries.reserve(entryCount);

	for (u32 index = 0; index < entryCount; index += 1) {
		const u8* entry = data + entryOffset + (index * entrySize);
		const u32 tokenLo = readLE32(entry + 0);
		const u32 tokenHi = readLE32(entry + 4);
		const AssetToken assetToken = makeAssetToken(tokenLo, tokenHi);
		const u32 typeId = readLE32(entry + 8);
		const u32 opId = readLE32(entry + 12);
		const u32 residOffset = readLE32(entry + 16);
		const u32 residLength = readLE32(entry + 20);
		const u32 sourceOffset = readLE32(entry + 24);
		const u32 sourceLength = readLE32(entry + 28);
		const u32 normalizedOffset = readLE32(entry + 32);
		const u32 normalizedLength = readLE32(entry + 36);
		const u32 updateLo = readLE32(entry + 80);
		const u32 updateHi = readLE32(entry + 84);

		const std::optional<std::string> assetId = decodeTocString(stringTable, stringTableSize, residOffset, residLength);
		if (!assetId.has_value()) {
			throw BMSX_RUNTIME_ERROR("ROM TOC entry missing asset id.");
		}
		if (hashAssetToken(*assetId) != assetToken) {
			throw BMSX_RUNTIME_ERROR("ROM TOC entry token mismatch for asset '" + *assetId + "'.");
		}

		RomAssetInfo romInfo;
		romInfo.type = assetTypeFromId(typeId);
		if (opId == ROM_TOC_OP_DELETE) {
			romInfo.op = std::string("delete");
		}
		romInfo.start = optionalI32FromU32(readLE32(entry + 40));
		romInfo.end = optionalI32FromU32(readLE32(entry + 44));
		romInfo.compiledStart = optionalI32FromU32(readLE32(entry + 48));
		romInfo.compiledEnd = optionalI32FromU32(readLE32(entry + 52));
		romInfo.metabufferStart = optionalI32FromU32(readLE32(entry + 56));
		romInfo.metabufferEnd = optionalI32FromU32(readLE32(entry + 60));
		romInfo.textureStart = optionalI32FromU32(readLE32(entry + 64));
		romInfo.textureEnd = optionalI32FromU32(readLE32(entry + 68));
		romInfo.collisionBinStart = optionalI32FromU32(readLE32(entry + 72));
		romInfo.collisionBinEnd = optionalI32FromU32(readLE32(entry + 76));
		romInfo.sourcePath = decodeTocString(stringTable, stringTableSize, sourceOffset, sourceLength);
		romInfo.normalizedSourcePath = decodeTocString(stringTable, stringTableSize, normalizedOffset, normalizedLength);
		const u64 updateTimestamp = (static_cast<u64>(updateHi) << 32) | updateLo;
		if (typeId == ROM_TOC_ASSET_TYPE_LUA || updateTimestamp != 0u) {
			romInfo.updateTimestamp = static_cast<i64>(updateTimestamp);
		}

		payload.entries.push_back(RomSourceEntry{*assetId, std::move(romInfo)});
	}
	return payload;
}

std::vector<u8> encodeRomToc(const RomTocPayload& payload) {
	std::vector<RomSourceEntry> entries = payload.entries;
	std::sort(entries.begin(), entries.end(), [](const RomSourceEntry& lhs, const RomSourceEntry& rhs) {
		return lhs.resid < rhs.resid;
	});

	std::vector<u8> stringTable;
	std::unordered_map<std::string, TocStringSlice> stringIndex;
	const TocStringSlice projectRoot = internTocString(stringTable, stringIndex, payload.projectRootPath);

	std::vector<u8> out(ROM_TOC_HEADER_SIZE + (entries.size() * ROM_TOC_ENTRY_SIZE));
	writeLE32(out.data() + 0, ROM_TOC_MAGIC);
	writeLE32(out.data() + 4, ROM_TOC_HEADER_SIZE);
	writeLE32(out.data() + 8, ROM_TOC_ENTRY_SIZE);
	writeLE32(out.data() + 12, static_cast<u32>(entries.size()));
	writeLE32(out.data() + 16, ROM_TOC_HEADER_SIZE);
	writeLE32(out.data() + 20, ROM_TOC_HEADER_SIZE + static_cast<u32>(entries.size() * ROM_TOC_ENTRY_SIZE));
	writeLE32(out.data() + 24, 0u);
	writeLE32(out.data() + 28, projectRoot.offset);
	writeLE32(out.data() + 32, projectRoot.length);
	writeLE32(out.data() + 36, 0u);
	writeLE32(out.data() + 40, 0u);
	writeLE32(out.data() + 44, 0u);

	for (size_t index = 0; index < entries.size(); ++index) {
		const RomSourceEntry& source = entries[index];
		const TocStringSlice resid = internTocString(stringTable, stringIndex, source.resid);
		const TocStringSlice sourcePath = internTocString(stringTable, stringIndex, source.rom.sourcePath);
		const TocStringSlice normalizedSourcePath = internTocString(stringTable, stringIndex, source.rom.normalizedSourcePath);
		const AssetTokenParts token = splitAssetToken(hashAssetToken(source.resid));
		const u32 opId = source.rom.op == "delete" ? ROM_TOC_OP_DELETE : ROM_TOC_OP_NONE;
		u8* entry = out.data() + ROM_TOC_HEADER_SIZE + (index * ROM_TOC_ENTRY_SIZE);
		writeLE32(entry + 0, token.lo);
		writeLE32(entry + 4, token.hi);
		writeLE32(entry + 8, assetTypeToId(source.rom.type));
		writeLE32(entry + 12, opId);
		writeLE32(entry + 16, resid.offset);
		writeLE32(entry + 20, resid.length);
		writeLE32(entry + 24, sourcePath.offset);
		writeLE32(entry + 28, sourcePath.length);
		writeLE32(entry + 32, normalizedSourcePath.offset);
		writeLE32(entry + 36, normalizedSourcePath.length);
		writeLE32(entry + 40, tocFieldValue(source.rom.start));
		writeLE32(entry + 44, tocFieldValue(source.rom.end));
		writeLE32(entry + 48, tocFieldValue(source.rom.compiledStart));
		writeLE32(entry + 52, tocFieldValue(source.rom.compiledEnd));
		writeLE32(entry + 56, tocFieldValue(source.rom.metabufferStart));
		writeLE32(entry + 60, tocFieldValue(source.rom.metabufferEnd));
		writeLE32(entry + 64, tocFieldValue(source.rom.textureStart));
		writeLE32(entry + 68, tocFieldValue(source.rom.textureEnd));
		writeLE32(entry + 72, tocFieldValue(source.rom.collisionBinStart));
		writeLE32(entry + 76, tocFieldValue(source.rom.collisionBinEnd));
		writeLE32(entry + 80, tocUpdateLo(source.rom.updateTimestamp));
		writeLE32(entry + 84, tocUpdateHi(source.rom.updateTimestamp));
	}

	writeLE32(out.data() + 24, static_cast<u32>(stringTable.size()));
	out.insert(out.end(), stringTable.begin(), stringTable.end());
	return out;
}

} // namespace bmsx

File: src/bmsx/rompack/toc.ts
import type { asset_type, RomAsset } from './format';

export const ROM_TOC_MAGIC = 0x434f5442; // 'BTOC' little-endian
export const ROM_TOC_HEADER_SIZE = 48;
export const ROM_TOC_ENTRY_SIZE = 88;
export const ROM_TOC_INVALID_U32 = 0xffffffff;
export const ROM_TOC_OP_NONE = 0;
export const ROM_TOC_OP_DELETE = 1;
export const ROM_TOC_ASSET_TYPE_IMAGE = 1;
export const ROM_TOC_ASSET_TYPE_AUDIO = 2;
export const ROM_TOC_ASSET_TYPE_DATA = 3;
export const ROM_TOC_ASSET_TYPE_BIN = 4;
export const ROM_TOC_ASSET_TYPE_ATLAS = 5;
export const ROM_TOC_ASSET_TYPE_ROMLABEL = 6;
export const ROM_TOC_ASSET_TYPE_MODEL = 7;
export const ROM_TOC_ASSET_TYPE_AEM = 8;
export const ROM_TOC_ASSET_TYPE_LUA = 9;
export const ROM_TOC_ASSET_TYPE_CODE = 10;
const utf8Decoder = new TextDecoder();

export type RomTocPayload = {
	entries: RomAsset[];
	projectRootPath: string | null;
};

export enum AssetTypeKind {
	ImageAtlas,
	Audio,
	Model,
	Aem,
	Bin,
	Lua,
	Data,
	Code,
	Skip,
	Unknown,
}

const ASSET_TYPE_IDS: Record<asset_type, number> = {
	image: ROM_TOC_ASSET_TYPE_IMAGE,
	audio: ROM_TOC_ASSET_TYPE_AUDIO,
	data: ROM_TOC_ASSET_TYPE_DATA,
	bin: ROM_TOC_ASSET_TYPE_BIN,
	atlas: ROM_TOC_ASSET_TYPE_ATLAS,
	romlabel: ROM_TOC_ASSET_TYPE_ROMLABEL,
	model: ROM_TOC_ASSET_TYPE_MODEL,
	aem: ROM_TOC_ASSET_TYPE_AEM,
	lua: ROM_TOC_ASSET_TYPE_LUA,
	code: ROM_TOC_ASSET_TYPE_CODE,
};

export function assetTypeToId(type: asset_type): number {
	const id = ASSET_TYPE_IDS[type];
	if (!id) {
		throw new Error(`Unknown asset type "${type}".`);
	}
	return id;
}

export function assetTypeFromId(id: number): asset_type {
	switch (id) {
		case ROM_TOC_ASSET_TYPE_IMAGE: return 'image';
		case ROM_TOC_ASSET_TYPE_AUDIO: return 'audio';
		case ROM_TOC_ASSET_TYPE_DATA: return 'data';
		case ROM_TOC_ASSET_TYPE_BIN: return 'bin';
		case ROM_TOC_ASSET_TYPE_ATLAS: return 'atlas';
		case ROM_TOC_ASSET_TYPE_ROMLABEL: return 'romlabel';
		case ROM_TOC_ASSET_TYPE_MODEL: return 'model';
		case ROM_TOC_ASSET_TYPE_AEM: return 'aem';
		case ROM_TOC_ASSET_TYPE_LUA: return 'lua';
		case ROM_TOC_ASSET_TYPE_CODE: return 'code';
		default:
			throw new Error(`Unknown asset type id "${id}".`);
	}
}

export function resolveAssetTypeKind(assetType: asset_type): AssetTypeKind {
	switch (assetType[0]) {
		case 'i':
			if (assetType === 'image') return AssetTypeKind.ImageAtlas;
			break;
		case 'a':
			if (assetType === 'atlas') return AssetTypeKind.ImageAtlas;
			if (assetType === 'audio') return AssetTypeKind.Audio;
			if (assetType === 'aem') return AssetTypeKind.Aem;
			break;
		case 'm':
			if (assetType === 'model') return AssetTypeKind.Model;
			break;
		case 'b':
			if (assetType === 'bin') return AssetTypeKind.Bin;
			break;
		case 'l':
			if (assetType === 'lua') return AssetTypeKind.Lua;
			break;
		case 'd':
			if (assetType === 'data') return AssetTypeKind.Data;
			break;
		case 'r':
			if (assetType === 'romlabel') return AssetTypeKind.Skip;
			break;
		case 'c':
			if (assetType === 'code') return AssetTypeKind.Code;
			break;
	}
	return AssetTypeKind.Unknown;
}

function decodeString(table: Uint8Array, offset: number, length: number, decoder: TextDecoder): string | null {
	if (offset === ROM_TOC_INVALID_U32 || length === 0) {
		return null;
	}
	return decoder.decode(table.subarray(offset, offset + length));
}

export function decodeRomToc(buffer: Uint8Array): RomTocPayload {
	if (buffer.byteLength < ROM_TOC_HEADER_SIZE) {
		throw new Error('ROM TOC buffer is too small.');
	}
	const view = new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength);
	const magic = view.getUint32(0, true);
	if (magic !== ROM_TOC_MAGIC) {
		throw new Error('Invalid ROM TOC magic.');
	}
	const headerSize = view.getUint32(4, true);
	if (headerSize !== ROM_TOC_HEADER_SIZE) {
		throw new Error(`Unexpected ROM TOC header size ${headerSize}.`);
	}
	const entrySize = view.getUint32(8, true);
	if (entrySize !== ROM_TOC_ENTRY_SIZE) {
		throw new Error(`Unexpected ROM TOC entry size ${entrySize}.`);
	}
	const entryCount = view.getUint32(12, true);
	const entryOffset = view.getUint32(16, true);
	const stringTableOffset = view.getUint32(20, true);
	const stringTableLength = view.getUint32(24, true);
	const projectRootOffset = view.getUint32(28, true);
	const projectRootLength = view.getUint32(32, true);
	if (entryOffset !== ROM_TOC_HEADER_SIZE) {
		throw new Error(`Unexpected ROM TOC entry offset ${entryOffset}.`);
	}
	const entriesByteLength = entryCount * entrySize;
	const expectedStringOffset = entryOffset + entriesByteLength;
	if (stringTableOffset !== expectedStringOffset) {
		throw new Error(`Unexpected ROM TOC string table offset ${stringTableOffset} (expected ${expectedStringOffset}).`);
	}

	const stringTable = buffer.subarray(stringTableOffset, stringTableOffset + stringTableLength);
	const projectRootPath = decodeString(stringTable, projectRootOffset, projectRootLength, utf8Decoder);

	const entries: RomAsset[] = [];
	for (let i = 0; i < entryCount; i += 1) {
		const base = entryOffset + i * entrySize;
		const tokenLo = view.getUint32(base + 0, true);
		const tokenHi = view.getUint32(base + 4, true);
		const typeId = view.getUint32(base + 8, true);
		const opId = view.getUint32(base + 12, true);
		const residOffset = view.getUint32(base + 16, true);
		const residLength = view.getUint32(base + 20, true);
		const sourceOffset = view.getUint32(base + 24, true);
		const sourceLength = view.getUint32(base + 28, true);
		const normalizedOffset = view.getUint32(base + 32, true);
		const normalizedLength = view.getUint32(base + 36, true);

		const start = view.getUint32(base + 40, true);
		const end = view.getUint32(base + 44, true);
		const compiledStart = view.getUint32(base + 48, true);
		const compiledEnd = view.getUint32(base + 52, true);
		const metaStart = view.getUint32(base + 56, true);
		const metaEnd = view.getUint32(base + 60, true);
		const textureStart = view.getUint32(base + 64, true);
		const textureEnd = view.getUint32(base + 68, true);
		const collisionBinStart = view.getUint32(base + 72, true);
		const collisionBinEnd = view.getUint32(base + 76, true);
		const updateLo = view.getUint32(base + 80, true);
		const updateHi = view.getUint32(base + 84, true);

		const resid = decodeString(stringTable, residOffset, residLength, utf8Decoder);
		if (!resid) {
			throw new Error('ROM TOC entry is missing resid.');
		}
		const entry: RomAsset = {
			resid,
			type: assetTypeFromId(typeId),
			id_token_lo: tokenLo,
			id_token_hi: tokenHi,
		};
		if (opId === ROM_TOC_OP_DELETE) {
			entry.op = 'delete';
		}
		const sourcePath = decodeString(stringTable, sourceOffset, sourceLength, utf8Decoder);
		const normalizedSourcePath = decodeString(stringTable, normalizedOffset, normalizedLength, utf8Decoder);
		if (sourcePath) entry.source_path = sourcePath;
		if (normalizedSourcePath) entry.normalized_source_path = normalizedSourcePath;

		if (start !== ROM_TOC_INVALID_U32) entry.start = start;
		if (end !== ROM_TOC_INVALID_U32) entry.end = end;
		if (compiledStart !== ROM_TOC_INVALID_U32) entry.compiled_start = compiledStart;
		if (compiledEnd !== ROM_TOC_INVALID_U32) entry.compiled_end = compiledEnd;
		if (metaStart !== ROM_TOC_INVALID_U32) entry.metabuffer_start = metaStart;
		if (metaEnd !== ROM_TOC_INVALID_U32) entry.metabuffer_end = metaEnd;
		if (textureStart !== ROM_TOC_INVALID_U32) entry.texture_start = textureStart;
		if (textureEnd !== ROM_TOC_INVALID_U32) entry.texture_end = textureEnd;
		if (collisionBinStart !== ROM_TOC_INVALID_U32) entry.collision_bin_start = collisionBinStart;
		if (collisionBinEnd !== ROM_TOC_INVALID_U32) entry.collision_bin_end = collisionBinEnd;

		const updateTimestamp = (updateHi * 0x100000000) + updateLo;
		if (updateTimestamp > 0) {
			entry.update_timestamp = updateTimestamp;
		}

		entries.push(entry);
	}

	return { entries, projectRootPath };
}
